torch/_dynamo/eval_frame.py

```diff
@@ -410,15 +408,8 @@ def __init__(
     ):
         def on_enter():
             global most_recent_backend
-            if OptimizeContext._different_backend(most_recent_backend, compiler_fn):
-                if config.raise_on_backend_change:
-                    raise ResetRequired()
-                else:
-                    warnings.warn(
-                        "changing options to `torch.compile()` may require "
-                        "calling `torch._dynamo.reset()` to take effect"
-                    )
             most_recent_backend = compiler_fn
+            guarded_backend_cache[id(most_recent_backend)] = most_recent_backend
```

Comment:
It seems like it would be better to have a thread local which is set when we enter a torch.compile region specifying what the current backend is. At the very least it needs to be thread local, which it doesn't seem like `most_recent_backend` is. (Yes, I know some of this is preexisting problems, but now you know.)

Reply:
I've modified accordingly and added a test for multi-threads but I'm not sure if the implementation is good enough.